### PR TITLE
Docs reorganisation

### LINKED ISSUE #1665: Docs reorganisation (again!)
Hi all,

I think the docs could really benefit from some simplification. A while ago, a decision was made to embrace a framework that included _background_, _tutorials_, _how-to-guides_ and _reference_. In retrosprect, I think this was premature and I think it's partially responsible for the docs feeling a little skeleton-y. The sidebar doesn't nicely match this framework.

I'd like to brainstorm opportunities for simplification here. One option for consideration is:

- Separate the docs very clearly (including the sidebar) into:
   - Overview
      - Discuss major features/modules (e.g. gamma, anaonymisation, etc.) of PyMedPhys and include links to user/contributors guides
         - _this enables a reader to find what they need according to feature, which is one natural path a reader is likely to want to take_
   - User guide
      - Library
      - App
      - CLI
   - Contributors guide
      - _TBD_

- Combine _tutorials_ and _how-tos_. I don't think it is useful to separate these until there is a marked different between them. The introduce clutter is not worth the added categorisation.

- Explore nesting of sidebar to improve organisation. If nesting is not possible, simply accept that not all pages need to be linked to sidebar.

-------------------------

Suggestions/thoughts? If there is enough support, I can make a PR work a live draft. But, although I have my own opinions, I'm very keen to hear from other PyMedPhys contributors/users - especially new ones.

### PR DESCRIPTION
**Pull request by Matthew-Jennings**
See #1665

You can see the changes made to the docs by this PR here: [https://docs.pymedphys.com/en/docs-reorg/](https://docs.pymedphys.com/en/docs-reorg/)

-------------------------------

Proposed structure (draft):

- Overview
- Users Guide
   - Getting started
      - Quick start guide
      - Using the Application
      - Using the CLI
      - Using the Library
   - How-to guides
   - Technical reference
   - Background information
- Contributors Guide
   - Getting started
      - Your first PR
      - How your PR will be reviewed
      - Joining the team!
   - Tips and tricks

### CODE REVIEW

**Comment by SimonBiggs** [comment]
https://jupyterbook.org/en/stable/intro.html is now being built with readthedocs which provides benefits such as handling multi-language and multi-version hosting. It also provides a way to handle the redirects discussed in the previous issue (GitHub pages doesn't provide an option for server-side redirects). As such I will set up a readthedocs install of our docs in preparation for this re-org.

**Comment by SimonBiggs** [comment]
Just a heads up, I am trying to get this branch to build over at:

https://pymedphys.readthedocs.io/en/docs-reorg/

**Comment by Matthew-Jennings** [comment]
Maybe I'll let the CLI finish then? There's some biggish changes in the latest commits

**Comment by SimonBiggs** [comment]
> Maybe I'll let the CLI finish then? There's some biggish changes in the latest commits

No need to wait for the CI to finish. https://pymedphys.readthedocs.io/en/docs-reorg/ builds on an independent server and is independent of the CI.

In fact, it's ready now! :)

**Comment by Matthew-Jennings** [comment]
Cool! So it's triggered on every commit? Every PR?

**Comment by Matthew-Jennings** [comment]
> It looks much, much cleaner 🙂, I really like it 😄

:heart:

**Comment by SimonBiggs** [comment]
> Cool! So it's triggered on every commit? Every PR?

Not every PR. We need to explicitly flag a branch for deployment. Generally it is used to be able to deploy multiple different versions at the same time from each of the version branches. But you can also make "hidden deployments" that can be used for checking.

File: lib/pymedphys/docs/release-notes.md
Comment (SimonBiggs) [inline lib/pymedphys/docs/release-notes.md:11]: We changed the reference file to instead be the README.rst file at the root of PyMedPhys. I find it by default makes much more sense for new contributors. I suspect you'll need to make sure you're editing the reference file (at the root directory)

Comment (Matthew-Jennings) [reply lib/pymedphys/docs/release-notes.md:11]: Ah yes, I remembered that for a different change (fixing a broken link) but forgot it here. Cheers.

**Comment by SimonBiggs** [comment]
By the way, I suspect there wouldn't be much harm in converting the following to a how-to (Speeding up gamma calculations):

https://pymedphys.readthedocs.io/en/docs-reorg/users/tutes/gamma/gamma_shell_accelerations.html

In hindsight, it is more of a how-to than a tutorial still...

**Comment by Matthew-Jennings** [comment]
> By the way, I suspect there wouldn't be much harm in converting the following to a how-to (Speeding up gamma calculations):
> 
> https://pymedphys.readthedocs.io/en/docs-reorg/users/tutes/gamma/gamma_shell_accelerations.html
> 
> In hindsight, it is more of a how-to than a tutorial still...

Yeah.

Honestly mate, even _after_ reading [the diataxis explanation](https://diataxis.fr/tutorials-how-to/), the difference between _tutorials_ and _how-to guides_ is still not clear to me. Would you mind fleshing out a little more precisely how you would differentiate them? Maybe with some examples that would clearly translate to our use case?

File: lib/pymedphys/docs/release-notes.md
Comment (Matthew-Jennings) [reply lib/pymedphys/docs/release-notes.md:11]: wait, I _did_ make this change in the CHANGELOG? Or did you?

**Comment by Matthew-Jennings** [comment]
> @Matthew-Jennings would you be able to make a https://readthedocs.org/ account and then I can add you as a maintainer over there?

Done. Signed up using Github.

**Comment by SimonBiggs** [comment]
> Maybe with some examples that would clearly translate to our use case?

Anything that is optimised for providing to a first time user of a part of PyMedPhys can be considered a tutorial. This is essentially a "getting started with module x.y.z" lession. Anything that helps a knowledgeable user do one specific task well that is often required (such as making gamma faster) is a how-to. The design goal is to have the how-to be short sharp, sweet, and to the point. Making it so that experienced users don't need to spend a lot of time parsing tutorial filler text. The design goal of the lesson is to be a great taster and first experience with a module, without ending up in the weeds.

A document can be written that is both a how-to and a tutorial. It just means that someone trying to use it as a tutorial will get confused with the hard concepts, and someone there for a how-to recipe will have to filter through all the hand holding to find the three to four lines of code they're looking for. The idea of trying to separate them is recognising that there is a direct audience partition. No one comes to a document wanting both a tutorial and a how-to. They generally only want one or the other. By blurring the lines the person coming for a tutorial can get readily confused, and the person coming for a how-to ends up with an inefficient document.

I think we remove the tutorial heading entirely, and then put a document under getting started > using the library > gamma. I suspect we could potentially partition the current gamma tutorial document into two separate documents. One being the how-to recipe. The other being the tutorial. The tutorial would be free to link to the how-to document saying "if you'd like to go into more detail about how to speed this up... etc click here".

**Comment by SimonBiggs** [comment]
> Done. Signed up using Github.

Do you happen to know your readthedocs username?

image

**Comment by Matthew-Jennings** [comment]
I think I've gotten as far as I can for now - _might_ be able to revisit later tonight. In summary

- I think the overall structure is basically ready. I don't believe there are broken links etc.
- Arguably the only glaring issue that probably need addressed prior to merge is that the [overview ](https://pymedphys.readthedocs.io/en/docs-reorg/overview.html) needs some serious fleshing out. I left some existing text there because it might still be useful, but the overview needs to describe the new structure and everything in it, of which the existing text is only a subsection.
- There are of course many other useful major and minor improvements that could be made, but I _think_ the issues they would improve upon are already present in the existing docs - so arguably need not prevent a merge (I'd still be intending upon creating new PRs to address some of these myself).

**Comment by Matthew-Jennings** [comment]
Oh, it's come through as "Centrus". Maybe I already had an account?! 😆

**Comment by SimonBiggs** [comment]
> * There are of course many other useful major and minor improvements that could be made, but I _think_ the issues they would improve upon are already present in the existing docs - so arguably need not prevent a merge (I'd still be intending upon creating new PRs to address some of these myself).

Yup, I agree. If it is strictly better than before then it is allowed to merge. Currently the only thing stopping that is the redirects. I'll make sure that docs.pymedphys.com points to the readthedocs version so that once we merge the redirects will be active.

**Comment by Matthew-Jennings** [comment]
> > * There are of course many other useful major and minor improvements that could be made, but I _think_ the issues they would improve upon are already present in the existing docs - so arguably need not prevent a merge (I'd still be intending upon creating new PRs to address some of these myself).
> 
> Yup, I agree. If it is strictly better than before then it is allowed to merge. Currently the only thing stopping that is the redirects. I'll make sure that docs.pymedphys.com points to the readthedocs version so that once we merge the redirects will be active.

Oh yes, I'd forgotten about the redirects, sorry.

**Comment by SimonBiggs** [comment]
Alright, we're now using readthedocs for https://docs.pymedphys.com/.

So, once we have the redirects sorted, this PR can now merge :slightly_smiling_face:. Beautiful stuff @Matthew-Jennings :slightly_smiling_face:

**Comment by SimonBiggs** [comment]
> Oh, it's come through as "Centrus". Maybe I already had an account?!

Added now :slightly_smiling_face:

**Comment by SimonBiggs** [comment]
Do we need a double up of the `ref` keyword here in the URL bar?

image

**Comment by SimonBiggs** [comment]
Hi @Matthew-Jennings, I've done the following redirects:

```
Page Redirect
/general/index.html -> /contrib/info/index.html
Page Redirect
/lib/ref/gamma.html -> /users/ref/lib-ref/gamma.html
Page Redirect
/lib/ref/index.html -> /users/ref/lib-ref/index.html
Page Redirect
/lib/ref/metersetmap.html -> /users/ref/lib-ref/metersetmap.html
Page Redirect
/lib/ref/mosaiq.html -> /users/ref/lib-ref/mosaiq.html
Page Redirect
/lib/ref/lib/trf.html -> /users/ref/lib-ref/trf.html
Page Redirect
/tutes/gamma/gamma_tutorial.html -> /users/tutes/gamma/gamma_shell_accelerations.html
Page Redirect
/lib/tutes/gamma/index.html -> /users/tutes/gamma/index.html
Page Redirect
/lib/tutes/index.html -> /users/tutes/index.html
Page Redirect
/general/open-source-benefits.html -> /contrib/why-open-source.html
Page Redirect
/contrib/other/index.html -> /contrib/tips-tricks/index.html
Page Redirect
/contrib/other/add-jupyter-kernel.html -> /contrib/tips-tricks/add-jupyter-kernel.html
Page Redirect
/general/file-structure.html -> /contrib/info/file-structure.html
Page Redirect
/contrib/documentation.html -> /contrib/info/documentation.html
Prefix Redirect
/ -> /en/latest/
```

There's still quite a few left though. Would you be able to take over from here?

**Comment by Matthew-Jennings** [comment]
> Do we need a double up of the `ref` keyword here in the URL bar?

No sorry, let's change

**Comment by SimonBiggs** [comment]
Also, while we're at it, I generally try to use shortened words as much as possible for URLs so as to keep them relatively small for people to copy-paste in locations.

One thing that might be worth a tweak is changing `tips-tricks` to instead just be `tips` or something like that?

**Comment by SimonBiggs** [comment]
Also, probably a much of a muchness but the name gamma_shell is more of an internal name. Potentially `gamma_shell_accelerations.html` could just be `gamma_accelerations.html` or `speed_up_gamma.html`. Or even just `speed_up.html` given it is already nested under the `gamma` directory...

**Comment by Matthew-Jennings** [comment]
> Also, while we're at it, I generally try to use shortened words for the directory structure that forms the URLs. This is so as to keep them relatively small for people to copy-paste in locations.
> 
> One thing that might be worth a tweak is changing `tips-tricks` to instead just be `tips` or something like that?

Yep, makes sense. The only reason I started having more descriptive names was that I was getting confused having multiple files open in VS code with the same name (the parent dir is not clearly visible - at least in my current config). But I think weblink convenience trumps that.

**Comment by Matthew-Jennings** [comment]
> Also, probably a much of a muchness but the name `gamma_shell` is more of an internal name, I wouldn't expect the `shell` to have much meaning to external users. Potentially `gamma_shell_accelerations.html` could just be `gamma_accelerations.html` or `speed_up_gamma.html`. Or even just `speed_up.html` given it is already nested under the `gamma` directory...

Sure =)

**Comment by Matthew-Jennings** [comment]
> There's still quite a few left though. Would you be able to take over from here?

All done! Assuming no errors...

I haven't put in any redirects for any of the "Tutorials"-specific pages, since none are present in the docs as they current stand in this PR. That might change in the future...

File: lib/pymedphys/docs/release-notes.md
Comment (SimonBiggs) [reply lib/pymedphys/docs/release-notes.md:11]: Yeah, I got confused I think. All good :slightly_smiling_face:

**Comment by SimonBiggs** [comment]
Beautiful :slightly_smiling_face:. I'll fix the propagate CI failure with the following approach:

https://github.com/pymedphys/pymedphys/issues/1681

**Comment by SimonBiggs** [comment]
Hopefully once https://github.com/pymedphys/pymedphys/pull/1682 merges in that can be used to update this branch and fix this broken CI :)

**Review by SimonBiggs — APPROVED** [review]
Beautiful stuff @Matthew-Jennings :slightly_smiling_face:

**Comment by SimonBiggs** [comment]
Just wanting to check @nlambriICH, are you okay with us moving your tutorial into the how-to section for now?

https://docs.pymedphys.com/en/docs-reorg/users/howto/gamma/speed-up.html

**Comment by nlambriICH** [comment]
Yup, no problem for me. Just a couple of comments:

1. The links to the "gamma" function are broken
2. Will the output of the first cell (the download status) be present in the final version of the docs? I find it really ugly 🙂

**Comment by Matthew-Jennings** [comment]
> 2\. Will the output of the first cell (the download status) be present in the final version of the docs? I find it really ugly

Curious. Those "download" outputs do not show when I build locally...

**Comment by SimonBiggs** [comment]
> Curious. Those "download" outputs do not show when I build locally...

One option is to trigger the downloads right after the following:

https://github.com/pymedphys/pymedphys/blob/abcc2784f0c2e48fbc579da9b8924d6741dfc77d/lib/pymedphys/_dev/docs.py#L51-L53

By running:

```python
FILES_TO_PRE_DOWNLOAD = ["original_dose_beam_4.dcm", "logfile_dose_beam_4.dcm"]

...

for file_name in FILES_TO_PRE_DOWNLOAD:
    pymedphys.data_path(file_name)
```

The thing is, those download prompts will only show up if those files haven't yet downloaded within the pymedphys data cache. I suspect locally on your machine they've already downloaded before. So no download prompts.

**Comment by Matthew-Jennings** [comment]
@SimonBiggs, the changes made from your [progagate simplifcation](https://github.com/pymedphys/pymedphys/commit/abcc2784f0c2e48fbc579da9b8924d6741dfc77d) have caused the docs to fail to re-build. Should we revert? Or do you think you'd be able to fix (I wouldn't know where to start at this point =/)?

**Comment by SimonBiggs** [comment]
Cheers for the heads up, that was the one big downside of readthedocs (miss-alignment between the CI and the final docs build process).

I wonder if we can use a readthedocs hook as a CI test...

**Comment by SimonBiggs** [comment]
Fixed :slightly_smiling_face:

**Comment by Matthew-Jennings** [comment]
> Cheers for the heads up, that was the one big downside of readthedocs (miss-alignment between the CI and the final docs build process).

No worries.

> I wonder if we can use a readthedocs hook as a CI test...

That would be cool!

**Comment by Matthew-Jennings** [comment]
I do not understand why, but the latest docs build no longer has the ugly downloads output. The other cells' outputs are visible.

**Comment by SimonBiggs** [comment]
> I do not understand why, but the latest docs build no longer has the ugly downloads output.

Potentially readthedocs caches the download? It would make sense from readthedocs point of view... potentially this is an unrelated issue to this PR and should be addressed separately.

I suspect, we should probably be pre-downloading within the docs command anyway. But that can be a separate PR. @nlambriICH  did you potentially want to implement in a separate PR what I detailed over at https://github.com/pymedphys/pymedphys/pull/1675#issuecomment-1158355751 ?

**Comment by SimonBiggs** [comment]
> Maaaybe merge time?

Yup! :tada: :slightly_smiling_face:

**Pull request merged**
